### ISSUE TITLE
Increasing minumum supported rust version occured by rkyv and hashbrown

### ISSUE THREAD

**Issue by brew0722** [wasmerio/wasmer#2819]
### Describe the bug
Close: https://github.com/CosmWasm/cosmwasm/issues/1244

Since the `rkyv` dependency referenced by several subpackages of wasmer is not pinned,
so the `rkyv` version is automatically incremented in the `cargo.lock`.
However, as the version of `rkyv` increases, the version of rust must be raised.
because of `Hashbrown 0.12.0` used in the latest version of rkyv requires rust `edition2021`.
So the original `wasmer 2.2.0` should be compilable with rust `1.54.0`,
but because of this problem, compiling elsewhere is not possible.
Until fix, we should present a guide to downgrade rkyv to new project with all dependencies.

```
user@AL01696733 burner % cargo +1.54.0 test   
error: failed to download `hashbrown v0.12.0`

Caused by:
  unable to get packages from source

Caused by:
  failed to parse manifest at `/Users/user/.cargo/registry/src/github.com-1ecc6299db9ec823/hashbrown-0.12.0/Cargo.toml`

Caused by:
  feature `edition2021` is required

  this Cargo does not support nightly features, but if you
  switch to nightly channel you can add
  `cargo-features = ["edition2021"]` to enable this feature
user@AL01696733 burner % 
```

```
    │   ├── wasmer-compiler v2.2.0
(...)
    │   │   ├── rkyv v0.7.36
(...)
    │   │   │   ├── hashbrown v0.12.0
```

So could you please pin the version of rkyv? Like wasmer's subpackages("=2.2.0).
https://github.com/wasmerio/wasmer/blob/8b9af9bee79c60ad4107ff58d36bf80ec76d3def/lib/compiler/Cargo.toml#L24
https://github.com/wasmerio/wasmer/blob/8b9af9bee79c60ad4107ff58d36bf80ec76d3def/lib/engine-dylib/Cargo.toml#L26
https://github.com/wasmerio/wasmer/blob/8b9af9bee79c60ad4107ff58d36bf80ec76d3def/lib/engine-universal/Cargo.toml#L27
https://github.com/wasmerio/wasmer/blob/8b9af9bee79c60ad4107ff58d36bf80ec76d3def/lib/types/Cargo.toml#L17
https://github.com/wasmerio/wasmer/blob/8b9af9bee79c60ad4107ff58d36bf80ec76d3def/lib/vm/Cargo.toml#L24

**Comment by syrusakbary** [comment]
Sounds reasonable!

**Comment by Amanieu** [comment]
Wasmer only guarantees support for the last 2 stable versions of Rust. You can see Wasmer's MSRV in the `rust-toolchain` file in the git repository, it is currently 1.56. In the near future we will be increasing the MSRV to 1.59 to use new language features (inline assembly).

**Comment by brew0722** [comment]
I understand your opinion.

Bu,t I think this is not just an inherent problem with wasmer 2.2.0. I have the same problem with wasmer 2.0.0.
wasmer 2.0.0 MSRV  is 1.51.0. Others using wasmer 2.0.0 will require msrv to be 1.55.0 or higher because of this issue.

Regardless of the current MSRV, if wasmer don't pin the dependencies version, it is expected that older versions in the future will continue to have similar problems.

**Comment by webmaster128** [comment]
Unfortunately this is an unresolved ecosystem-wide problem. I highly recommend taking the time to read all of https://github.com/rust-lang/api-guidelines/issues/252, which was very helpful to me. It's a back and forth between _You can't simply break other people's projects that compiled once with compiler X_ on the one side and _well, stability guarantess? Whatever! Give it up because there is always this one dude in the dependency tree that requires a recent Rust version for some reason if any_.

CosmWasm tries to establish compiler support guarantees and we currently cannot easily upgrade beyond 1.55.0 for [very custom reasons](https://github.com/CosmWasm/wasmvm/issues/277). But we are failing for the reasons described above.

**Comment by jcaesar** [comment]
I don't think wasmer should fix a maximum version for rkyv. Suppose rkyv releases some important bugfix. If wasmer requires a specific version,  dependencies of wasmer wouldn't be able to use the rkyv bugfix release without an update from wasmer first.

Would adding a dependency like
```
rkyv = { version = "<0.7.36" }
```
to cosmwasm take care of the problem?

**Comment by webmaster128** [comment]
I agree there is very little Wasmer can or should do about it. It's a Rust ecosytsem mess and we have the same problem potentially for each and every dependency. We see many projects even bumping their compiler requirements in a patch release.

I work around it by manually adjusting lockfiles. For our users there is little we can do to provide a stable min compiler version.

**Comment by brew0722** [comment]
I agree that this is essentially a problem on the rust eco system side.
And I also understand that it is some W/A to pin rkyv.

However,Solving with W/A in the cosmwasm project is not completely solved the problem yet.
This is because the cargo.lock of the newly created contract package cannot be pinned rkyv by cosmwasm package.
After all, we should always guide contract developers (wasm binary developers) how to pin the rkyv.

Anyway, I understand that it is difficult for wasmer to guarantee MSRV. Please close if there are no other comments.

**Comment by syrusakbary** [comment]
It seems we should be good to close the issue, since there's little we can do on our side.

But if anyone else has more suggestions on what can be done on our side please feel free to add a comment
